### ISSUE TITLE
Missed optimization of tautological assignment of a global variable

### ISSUE THREAD

**Issue by artem** [llvm/llvm-project#63419]
Consider the following code:

```c++
unsigned long long value;

void foo() {
    if (value == 4){
        value = 4; 
    }
}
```
https://godbolt.org/z/azYWWboba

Clang doesn't detect a tautological assignment, which can be optimized out.

**Comment by changkhothuychung** [comment]
I would like to take this issue, if there can be some instructions how to get started since I am still new to llvm. Thanks!

**Comment by changkhothuychung** [comment]
@dtcxzyw dt
hi thanks for assigning this to me. Can you let me know which files I can look into, and some instructions to get started with this?

**Comment by nikic** [comment]
If you're new to LLVM, I probably wouldn't recommend working on this issue.

The place to implement this would be DeadStoreElimination.cpp, and the general thing to loop for is a store with a dominating condition that has an icmp with a load from the same value, and then check whether the clobbering access of the store dominates the load access.

The trick bit is the "with a dominating condition" bit. You could easily do this if you only support the condition on the immediate dominator. Otherwise you would probably want to walk the dom-tree and remember which relevant dominating conditions you currently know about.

**Comment by changkhothuychung** [comment]
> If you're new to LLVM, I probably wouldn't recommend working on this issue.
> 
> The place to implement this would be DeadStoreElimination.cpp, and the general thing to look for is a store with a dominating condition that has an icmp with a load from the same value, and then check whether the clobbering access of the store dominates the load access.
> 
> The trick bit is the "with a dominating condition" bit. You could easily do this if you only support the condition on the immediate dominator. Otherwise you would probably want to walk the dom-tree and remember which relevant dominating conditions you currently know about.

Thanks! Feel free to unassign me and assign it someone else who is interested in this topic and has more knowledge than me. I will jus try my best to see what I can do. 

However, do you have any recommendation on exsting issues that do similar work but easier? Maybe I can try those first and come back to this one later. Any input is apprciated!

**Comment by BK1603** [comment]
Hi @nikic, since changkhothuychung decided not to work on this, I created a PR addressing this issue (#75744). I've tested the change with opt but was not able to generate a test file for this. (Shared the issue that I am running into on PR.) Can you please help me with that?
